### ISSUE TITLE
Segmentation fault with dorado 0.7 basecaller and barcode kit SQK-NBD114-96

### ISSUE THREAD

**Issue by mprous1** [nanoporetech/dorado#860]
# Issue Report

## Please describe the issue:

For one small dataset (3.5 Gb pod5 files), dorado 0.7 with --kit-name SQK-NBD114-96 fails to finish basecalling (reaches 80% or so) and leaves truncated bam file. Resuming from this fail does not work. When taking a subset of the data (pod5 pass folders about 2.8 Gb), bascalling reaches 96% and basecalling the remaining pod5 folders separately works fine.

However, it finishes all data without --kit-name option.

At the same time, dorado 0.5 works fine with --kit-name option for the whole dataset.

Also, dorado 0.7 finishes basecalling of a different similarly sized dataset (4 Gb pod5) with --kit-name SQK-NBD114-96.

One other issue I noticed is that dorado 0.7 is maybe about 10 times slower compared to 0.5 for the same dataset. Is it because of the transformer model? 10-15 min for dorado 0.5, 1.5-2 h for dorado 0.7. Is that big difference normal?

## Run environment:
- Dorado version: 0.7.0+71cc744
- Dorado command: dorado basecaller sup pod5 --kit-name SQK-NBD114-96 -r > reads.bam
- or same result with Dorado command: dorado basecaller .temp_dorado_model-309c532b00029178/dna_r10.4.1_e8.2_400bps_sup@v5.0.0 pod5 --kit-name SQK-NBD114-96 -r > reads.bam
- Operating system: Kernel: 5.15.0-107-generic x86_64 bits: 64 compiler: gcc v: 11.4.0 Desktop: Cinnamon 6.0.4
    tk: GTK 3.24.33 wm: muffin vt: 7 dm: LightDM 1.30.0 Distro: Linux Mint 21.3 Virginia
    base: Ubuntu 22.04 jammy

- Hardware (CPUs, Memory, GPUs): 24-core (8-mt/16-st) model: Intel Core i9-14900KF bits: 64 type, 128 Gb RAM, NVIDIA GeForce RTX 4090/PCIe/SSE2 v: 4.6.0 NVIDIA 535.171.04

- Source data type : pod5
- Source data location: on device
- Details about data: SQK-NBD114-96, short reads (most around 110-400 bp), about 1.1 M reads, 3.5 Gb of pod5
- Dataset to reproduce, if applicable (small subset of data to share as a pod5 to reproduce the issue): I guess I could try to send 2.8 Gb of pod5 data, I would not know which subset to take to reproduce the issue.

## Logs
[2024-06-01 08:48:24.392] [info] Running: "basecaller" ".temp_dorado_model-309c532b00029178/dna_r10.4.1_e8.2_400bps_sup@v5.0.0" "pod5" "--kit-name" "SQK-NBD114-96" "-r"
[2024-06-01 08:48:24.414] [info] > Creating basecall pipeline
[2024-06-01 08:48:29.385] [info] cuda:0 using chunk size 12288, batch size 128
[2024-06-01 08:48:29.760] [info] cuda:0 using chunk size 6144, batch size 128
Segmentation fault (core dumped) 81% [01h:24m:20s<18m:47s] Basecalling

**Comment by dhoconno** [comment]
I'm seeing the same error:

[2024-06-03 16:02:18.611] [info] Running: "basecaller" "--device" "cuda:0" "--kit-name" "SQK-NBD114-96" "sup" "pod5"
[2024-06-03 16:02:38.156] [info] Assuming cert location is /etc/ssl/certs/ca-bundle.crt
[2024-06-03 16:02:38.164] [info]  - downloading dna_r10.4.1_e8.2_400bps_sup@v5.0.0 with httplib
[2024-06-03 16:02:41.932] [info] > Creating basecall pipeline
[2024-06-03 16:03:08.857] [info] cuda:0 using chunk size 12288, batch size 160
[2024-06-03 16:03:09.454] [info] cuda:0 using chunk size 6144, batch size 160
/var/lib/condor/execute/slot1/dir_1607271/30236-basecall.sh: line 26:     8 Segmentation fault      (core dumped) dorado-0.7.0-linux-x64/bin/dorado basecaller --device cuda:0 --kit-name SQK-NBD114-96 sup pod5 > unmapped.bam

**Comment by DNA-Dave** [comment]
I'm also seeing the same error:

[2024-06-03 20:01:46.866] [info] Running: "basecaller" "-r" "/home/wud3/repos/dorado-0.7.0-linux-x64/models/dna_r10.4.1_e8.2_400bps_sup@v5.0.0" "/mnt/isilon/lin_lab_share/BDB_PBMC/Bulk_1D_BDB03" "--kit-name" "SQK-NBD114-24"
[2024-06-03 20:01:46.993] [info] > Creating basecall pipeline
[2024-06-03 20:02:04.202] [info] cuda:0 using chunk size 12288, batch size 512
[2024-06-03 20:02:04.304] [info] cuda:0 using chunk size 6144, batch size 512
dorado.sh: line 4: 1394197 Segmentation fault      (core dumped) dorado basecaller -r /home/wud3/repos/dorado-0.7.0-linux-x64/models/dna_r10.4.1_e8.2_400bps_sup@v5.0.0 /mnt/isilon/lin_lab_share/BDB_PBMC/$1 --kit-name SQK-NBD114-24 > $1_basecalled.bam

**Comment by tijyojwad** [comment]
Hi all, apologies for the late response. This clearly looks like a bug. But since it's a segfault it can be hard to reproduce. @DNA-Dave @dhoconno are your reads short as well?

If anyone has been able to reproduce it on a small dataset, that would be ideal.

> dorado 0.7 is maybe about 10 times slower compared to 0.5 for the same dataset

yes the v5 SUP model is the new architecture with much higher accuracy. But it's not fully optimized yet and we're actively working on it. So in this release it will run slower than v0.4.3 SUP model.

**Comment by DNA-Dave** [comment]
Hi @tijyojwad I'm using dorado to basecall full length cDNA reads for long read RNA-seq. If I had to guess, mean read length should be around 1-2kb

**Comment by dhoconno** [comment]
Yes, my reads are short. Most are going to be <500bp.

**Comment by jts** [comment]
I have this issue as well, on short DNA

**Comment by jts** [comment]
A few notes on my case:

-dorado 0.7.1 with hac@v4.3.0 also seg faults, dorado 0.6.0 works
-removing `--kit-name` works as for @mprous1
-using `--kit-name` but setting `--trim none` works

**Comment by jts** [comment]
Here's a pod5 file with a single read to reproduce the issue:


[barcode_issue.tar.gz](https://github.com/user-attachments/files/15787850/barcode_issue.tar.gz)

```
dorado basecaller dna_r10.4.1_e8.2_400bps_hac@v4.3.0 barcode_issue.pod5 --kit-name EXP-NBD196 --device cuda:all 
```

Curiously only the linux build seems affected, it does not seg fault on MacOS.

**Comment by mprous1** [comment]
Thanks @jts , I was about start testing which of my pod5 files/folders could be representative for this.

**Comment by malton-ont** [comment]
Hi @jts, @mprous1,

We've been able to reproduce the segfault, and we've found an issue where short reads may be identified as being entirely barcode and are completely trimmed away, leading to unexpected behaviour downstream. We'll get a fix out asap.

Thanks again for reporting this!

**Comment by mprous1** [comment]
Great that this got resolved!

**Comment by DNA-Dave** [comment]
I've played around with things and found that a pod5 file that segfaults with:

dorado basecaller -r ***MODEL_PATH*** ***POD5_PATH*** --kit-name SQK-NBD114-24 > $1_basecalled.bam

runs without issue with:

dorado basecaller -r ***MODEL_PATH*** ***POD5_PATH*** | dorado demux --kit-name SQK-NBD114-24 --output-dir ***OUTPUT_PATH***

My question is: these two commands should produce exactly equivalent outputs, correct?

**Comment by malton-ont** [comment]
Hi @DNA-Dave,

The results won't be exactly the same. For one, the second command will output the reads into separate files based on the barcode. It will also leave all reads with the same read group tag (`RG`), while the first command splits different barcodes into different read groups. The actual classifications should be the same though.

**Comment by jts** [comment]
Sorry to bug @malton-ont but is there a timeline for this fix to be released? I have some data I need to basecall and am trying to decide whether to wait or downgrade to 0.6

**Comment by malton-ont** [comment]
@jts Hopefully in the next week, but no promises!

**Comment by jts** [comment]
Ok understood, thanks!

**Comment by iiSeymour** [comment]
@jts we have one more bug fix before cutting v0.7.2 so probably next week as @malton-ont says. I can sync the release branch now if that helps.

**Comment by jts** [comment]
yeah that would be great, thanks @iiSeymour and @malton-ont

**Comment by iiSeymour** [comment]
@jts done - head of the branch _(release-v0.7)_ has the fix.

**Comment by iiSeymour** [comment]
Fixed in v0.7.2 release just now.
